File: LeetCode/101. Symmetric Tree/101.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        q.push(root);
        
        while(!q.empty()) {
            TreeNode* t1 = q.front();
            q.pop();
            TreeNode* t2 = q.front();
            q.pop();
            
            if(t1 == NULL && t2 == NULL) continue;
            if(t1 == NULL || t2 == NULL) return false;
            if(t1->val != t2->val) return false;
            
            q.push(t1->left);
            q.push(t2->right);
            q.push(t1->right);
            q.push(t2->left);
        }
        
        return true;
        
    }
};

//other people's solution
/*
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == nullptr) return true;
        return subMirror(root->left,root->right);
    }
    bool subMirror(TreeNode* left, TreeNode* right){
        if(not left and not right){
            return true;
        } else if(not left or not right or left->val != right->val){
            return false;
        }
        return subMirror(left->left, right->right) and subMirror(right->left, left->right);
    }
};
*/

/*
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return (!root) ? true : isSS(root -> left, root -> right);
    }

    bool isSS(TreeNode* l, TreeNode* r) {
        if (!l || !r)
            return l == r;
        return (l -> val == r -> val) && isSS(l -> left, r -> right) && isSS(l -> right, r -> left);
    }
};
*/
/*
class Solution {
public:
bool isMirror(TreeNode* left, TreeNode* right) {
    if (left == NULL && right == NULL) return true;
    if (left == NULL || right == NULL) return false;
    return (left->val == right->val) && isMirror(left->left, right->right) 
                                     && isMirror(left->right, right->left);
}

    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
    return isMirror(root->left, root->right);
    }
};
*/
/*
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        queue<TreeNode*> q;
        if(root==nullptr)return true;
        if(root->left==nullptr && root->right==nullptr)return true;
        q.push(root);
        while(!q.empty()){
            vector<TreeNode*> level;
            int size=q.size();
            for(int i=0;i<size;i++){
                TreeNode* node=q.front();
                q.pop();
                level.push_back(node);
                if(node!=nullptr)q.push(node->left);
                if(node!=nullptr)q.push(node->right);
            }
            // cout <<level.size()<<endl;
            // if(level.size()%2!=0){
            //     return false;
            // }
            for(int i=0;i<level.size()/2;i++){
                if(level[i]==nullptr && level[level.size()-i-1]!=nullptr)return false;
                if(level[i]!=nullptr && level[level.size()-i-1]==nullptr)return false;
             
                if(level[i]!=nullptr ){
                    if(level[i]->val!=level[level.size()-1-i]->val)return false;
                }
                else{
                    if(level[i]!=level[level.size()-1-i])return false;
                }
            
            }
        }
        return true;

    }
};
*/
/*
class Solution {
public:
    bool isSymmetric(TreeNode* root) 
    {
        queue<pair<TreeNode*,TreeNode*>>a;
        bool isSym = true;
        a.push(make_pair(root->left, root->right));
        while(!a.empty())
        {
            auto i = a.front();
            if ((i.first && !i.second) || (!i.first && i.second))
            {
                isSym = false;
                break;
            }
            if (i.first && i.second)
            {
                if (i.first->val != i.second->val)
                {
                    isSym = false;
                    break;
                }
                a.push(make_pair(i.first->left,i.second->right));
                a.push(make_pair(i.first->right,i.second->left));
            }
            a.pop();
        }
        return isSym;
    }
};
*/

File: LeetCode/101. Symmetric Tree/101.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        // if root is null, return true
        return IsMirror(root, root);
        
    }
    public bool IsMirror(TreeNode t1, TreeNode t2){
        // if both are null, return true
        if(t1 == null && t2 == null) return true;
        // if one of them is null, return false
        if(t1 == null || t2 == null) return false;
        // if both are not null, check if their values are equal and their children are mirror
        return (t1.val == t2.val) && IsMirror(t1.right, t2.left) && IsMirror(t1.left, t2.right);
    }
}
//other people's solution

/*
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        if(root == null)
            return true;
        if(root.left == null || root.right == null)
            return(root.left == null && root.right == null);

        return(CheckLeftToRight(root.left, root.right));       
    }

    public bool CheckLeftToRight(TreeNode left, TreeNode right){
        if(left == null || right == null)
            return(left == null && right == null);
        if(left.val != right.val)
            return false;
        return(CheckLeftToRight(left.left, right.right) && CheckLeftToRight(left.right, right.left));
    }
}
*/
/*
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        return Traverse(root.left, root.right);
    }
    
    bool Traverse(TreeNode left, TreeNode right)
    {
        if(left is null && right is null)
            return true;
        
        if(left is null || right is null || left.val != right.val)
            return false;
        
        var leftEq = Traverse(left.left, right.right);
        
        if(!leftEq)
            return false;
        
        var rightEq = Traverse(left.right, right.left);
        
        return rightEq;
    }
}
*/

/*
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        if(root == null){
            return true;
        }
        else{
            if(root.left == root.right){
                return true;
            }
                return taraflar(root.left, root.right);
        }

    }
    private bool taraflar(TreeNode left , TreeNode right) {
        if (left == null && right == null)
        {
            return true;
        }

        if (left == null || right == null)
        {
            return false;
        }

        return (left.val == right.val) &&
               taraflar(left.left, right.right) &&
               taraflar(left.right, right.left);
    }
}
*/